internal/graph/rate_limited_check_resolver.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// RateLimitedCheckResolver will prioritize requests with fewer dispatches over
+// requests with more dispatches.
+// Initially, request's dispatches will not be throttled and will be processed
+// immediately. As the number of dispatches increases (to beyond the config's MediumPriorityLevel),
+// selectively number (configure via config's MediumPriorityShaper) will be throttled by placing the
+// dispatch in the medium priority queue. One item from the medium priority queue will be processed
+// every ticker across the entire server. Request's dispatches will gradually be more frequently throttled as
+// the number of dispatches increase until every dispatch will be throttled by being placed in the medium priority
+// queue. When the number of request dispatches is above the LowPriorityLevel, the dispatches are placed
+// in the low priority queue. One item form the low priority queue will be processed every Nth ticker
+// (configured via config's LowPriorityShaper parameter).
+// This allows a check / list objects request to be gradually throttled.
+type RateLimitedCheckResolver struct {
+	delegate         CheckResolver
+	config           RateLimitedCheckResolverConfig
+	ticker           *time.Ticker
+	lowPriorityQueue chan bool
+	medPriorityQueue chan bool
+	done             chan bool
+
+	// these are helper config value to reduce calculation needed for medium shaper
+	mediumPriorityLevel2      uint32
+	mediumPriorityLevel3      uint32
+	mediumPriorityLevel4      uint32
+	mediumPriorityShaperFreq2 uint32
+	mediumPriorityShaperFreq3 uint32
+}
+
+var _ CheckResolver = (*RateLimitedCheckResolver)(nil)
+
+var (
+	rateLimitedCheckResolverDelayMsHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace:                       build.ProjectName,
+		Name:                            "rate_limited_check_resolver_delay_ms",
+		Help:                            "Time spent waiting for rate limited check resolver",
+		Buckets:                         []float64{1, 3, 5, 10, 25, 50, 100, 1000, 5000}, // Milliseconds. Upper bound is config.UpstreamTimeout.
+		NativeHistogramBucketFactor:     1.1,
+		NativeHistogramMaxBucketNumber:  100,
+		NativeHistogramMinResetDuration: time.Hour,
+	}, []string{"grpc_service", "grpc_method", "queue_name"})
+)
+
+func NewRateLimitedCheckResolver(
+	config RateLimitedCheckResolverConfig) *RateLimitedCheckResolver {
+	ticker := time.NewTicker(config.TimerTickerFrequency)
+	rateLimitedCheckResolver := &RateLimitedCheckResolver{
+		config:           config,
+		ticker:           ticker,
+		lowPriorityQueue: make(chan bool),
+		medPriorityQueue: make(chan bool),
+		done:             make(chan bool),
+	}
+	rateLimitedCheckResolver.setMediumConfig(config)
+	rateLimitedCheckResolver.delegate = rateLimitedCheckResolver
+	go rateLimitedCheckResolver.runTicker()
+	return rateLimitedCheckResolver
+}
+
+func (r *RateLimitedCheckResolver) setMediumConfig(config RateLimitedCheckResolverConfig) {
```

Comment:
So, it looks like this function is taking the ranges

[0, mediumlevel, lowlevel]

and further dividing them into

[0, mediumlevel2, mediumlevel3, mediumlevel4, lowlevel]

is my understanding correct? if so can you add godocs for it? please also explain _why_ we need to further divide it

Reply:
comment obsolete as we only have 1 queue